addons/sms/static/tests/web/sms_button.test.js

```diff
@@ -0,0 +1,142 @@
+import { mockService, mountView } from "@web/../tests/web_test_helpers";
+
+describe.current.tags("desktop");
+defineSMSModels();
+
+beforeEach(async () => {
+    const pyEnv = await startServer();
+    pyEnv["partner"].create([
+        { message: "", foo: "yop", mobile: "+32494444444"},
+        { message: "", foo: "bayou"},
+    ]);
+    pyEnv["visitor"].create([
+        { mobile: "+32494444444" },
+    ]);
+})
+
+test("Sms button in form view", async () => {
+    await mountView({
+        type: "form",
+        resModel: "visitor",
+        resId: 1,
+        mode: "readonly",
+        arch:
+            `<form>
+                <sheet>
+                    <field name="mobile" widget="phone"/>
+                </sheet>
+            </form>`
+    });
+    await contains(".o_field_phone");
+    await contains(".o_field_phone a.o_field_phone_sms", { count: 1 });
+});
+
+test("Sms button with option enable_sms set as False", async () => {
+    await mountView({
+        type: "form",
+        resModel: "visitor",
+        resId: 1,
+        mode: "readonly",
+        arch:
+            `<form>
+                <sheet>
+                    <field name="mobile" widget="phone" options="{'enable_sms': false}"/>
+                </sheet>
+            </form>`
+    });
+    await contains(".o_field_phone");
+    await contains(".o_field_phone a.o_field_phone_sms", { count: 0 });
+});
+
+test("click on the sms button while creating a new record in a FormView", async () => {
+    mockService("action", {
+        doAction(action, options) {
+            step("do_action");
+            expect(action.type).toBe("ir.actions.act_window");
+            expect(action.res_model).toBe("sms.composer");
+            options.onClose();
+        },
+    });
+    await mountView({
+        type: "form",
+        resModel: "partner",
+        resId: 1,
+        arch:
+            `<form>
+                <sheet>
+                    <field name="foo"/>
+                    <field name="mobile" widget="phone"/>
+                </sheet>
+            </form>`,
+    });
+    await editInput(document.body, "[name='foo'] input", "John");
+    await editInput(document.body, "[name='mobile'] input", "+32494444411");
+    await click(".o_field_phone_sms");
+    expect(queryFirst("[name='foo'] input")).toHaveValue("John");
+    expect(queryFirst("[name='mobile'] input")).toHaveValue("+32494444411");
+    await assertSteps(["do_action"]);
+});
+
+
+test(
+    "click on the sms button in a FormViewDialog has no effect on the main form view",
+    async () => {
+        mockService("action", {
+            doAction(action, options){
+                step("do_action");
+                expect(action.type).toBe("ir.actions.act_window");
+                expect(action.res_model).toBe("sms.composer");
+                options.onClose();
+            },
+        });
+        await mountView({
+            type: "form",
+            resModel: "partner",
+            resId: 1,
```

Comment:
This is fragile as you are asserting specific partner data later. You should instead retrieve the partner_id from the server and use it instead of this hard coded "1".

e.g.

```js
partnerId = pyEnv["res.partner"].search([["foo", "=", "yop"]]);
```

Reply:
@tsm-odoo Changes done. Please have a look.